DICOM reader: better handling of nested sequences that include pixel data

melissalinkert opened the pull request:
Fixes #4165.

Using the data in `inbox/gh-4165`, `showinf` should show that each pyramid resolution starts with "scrambled" tiles. `showinf -noflat -resolution 2 img_0.dcm` or `showinf -nogroup img_3.dcm` in particular can easily reproduce the issue.

With this change, each resolution and extra image in the dataset should display correctly.

As described in #4165, there are PixelData elements (which Bio-Formats' reader should ignore) that are nested within a private sequence element. The reader will no longer stop at the first PixelData encountered, and instead should skip over any PixelData contained within a sequence.

melissalinkert committed:
DICOM reader: better handling of nested sequences that include pixel data

Fixes #4165.

sbesson reviewed (commented):
Tested using the public DICOM samples shared in #4165. Without this PR I confirm the whole slide image appears as scrambled as the result of the Pixel Data handling.
<img width="2160" alt="Screenshot 2024-03-29 at 12 06 41" src="https://github.com/ome/bioformats/assets/1355463/51b74224-4d75-4cca-afa3-31ca77b825ce">

With this PR included, all resolution levels of the whole slide image as well as the associated image read without issue using the command-line utility.

One minor concern looking the metadata output of the `showinf -noflat` and comparing it with the output of the `dcdump` utility on the same file is that the series metadata includes information for all resolutions of the hidden pyramid.

```
% ./tools/showinf -noflat -nopix ~/Downloads/gh-4165/DCM/img_0.dcm | grep Total
0048,0006 Total Pixel Matrix Columns #1: 46000
0048,0006 Total Pixel Matrix Columns #2: 23000
0048,0006 Total Pixel Matrix Columns #3: 11500
0048,0006 Total Pixel Matrix Columns #4: 5750
0048,0006 Total Pixel Matrix Columns #5: 2875
0048,0006 Total Pixel Matrix Columns #6: 1437
0048,0006 Total Pixel Matrix Columns #7: 718
0048,0006 Total Pixel Matrix Columns #8: 359
0048,0006 Total Pixel Matrix Columns #9: 179
0048,0007 Total Pixel Matrix Rows #1: 32914
0048,0007 Total Pixel Matrix Rows #2: 16457
0048,0007 Total Pixel Matrix Rows #3: 8228
0048,0007 Total Pixel Matrix Rows #4: 4114
0048,0007 Total Pixel Matrix Rows #5: 2057
0048,0007 Total Pixel Matrix Rows #6: 1028
0048,0007 Total Pixel Matrix Rows #7: 514
0048,0007 Total Pixel Matrix Rows #8: 257
0048,0007 Total Pixel Matrix Rows #9: 128
0048,0303 Total Pixel Matrix Focal Planes: 1
```

This information is slightly at odds with the DICOM resolutions detected by the reader which only include 3 levels of sizeX 46000 11500 and 2875.

melissalinkert committed:
Exclude private content sequences from the original metadata table

The `getTags()` method still includes these sequences, but this just
keeps the original metadata table a little easier to read.
Private content sequences are detected by looking for a
private content creator, then excluding any sequences with a matching
high word. This corresponds to https://dicom.nema.org/dicom/2013/output/chtml/part05/sect_7.8.html

melissalinkert commented:
daaa2f9 should fix the original metadata table issue in https://github.com/ome/bioformats/pull/4168#pullrequestreview-1968481092

sbesson reviewed (approved):
With the last commit, the difference in the original metadata for the first series of the sample shared in #4165 is

 ```diff
8,9c8
< 0008,0008 Image Type #1: DERIVED\PRIMARY\VOLUME\NONE
< 0008,0008 Image Type #2: DERIVED\PRIMARY\VOLUME\RESAMPLED
---
> 0008,0008 Image Type: DERIVED\PRIMARY\VOLUME\NONE
87,88d85
< 0008,1150 Referenced SOP Class UID: 1.2.840.10008.5.1.4.1.1.77.1.6
< 0008,1155 Referenced SOP Instance UID: 1.3.6.1.4.1.5962.99.1.3527156866.2038118173.1708629206146.9.0
104,105c101
< 0020,000e Series Instance UID #1: 1.3.6.1.4.1.5962.99.1.3527156866.2038118173.1708629206146.4.0
< 0020,000e Series Instance UID #2: 1.3.6.1.4.1.5962.99.1.3527156866.2038118173.1708629206146.4.0
---
> 0020,000e Series Instance UID: 1.3.6.1.4.1.5962.99.1.3527156866.2038118173.1708629206146.4.0
119,130c115,120
< 0028,0002 Samples per pixel #1: 3
< 0028,0002 Samples per pixel #2: 3
< 0028,0004 Photometric Interpretation #1: RGB
< 0028,0004 Photometric Interpretation #2: YBR_FULL_422
< 0028,0006 Planar Configuration #1: 0
< 0028,0006 Planar Configuration #2: 0
< 0028,0008 Number of frames #1: 23220
< 0028,0008 Number of frames #2: 5850
< 0028,0010 Rows #1: 256
< 0028,0010 Rows #2: 256
< 0028,0011 Columns #1: 256
< 0028,0011 Columns #2: 256
---
> 0028,0002 Samples per pixel: 3
> 0028,0004 Photometric Interpretation: RGB
> 0028,0006 Planar Configuration: 0
> 0028,0008 Number of frames: 23220
> 0028,0010 Rows: 256
> 0028,0011 Columns: 256
132,137c122,124
< 0028,0100 Bits Allocated #1: 8
< 0028,0100 Bits Allocated #2: 8
< 0028,0101 Bits Stored #1: 8
< 0028,0101 Bits Stored #2: 8
< 0028,0102 High Bit #1: 7
< 0028,0102 High Bit #2: 7
---
> 0028,0100 Bits Allocated: 8
> 0028,0101 Bits Stored: 8
> 0028,0102 High Bit: 7
159,162c146,147
< 0048,0006 Total Pixel Matrix Columns #1: 46000
< 0048,0006 Total Pixel Matrix Columns #2: 23000
< 0048,0007 Total Pixel Matrix Rows #1: 32914
< 0048,0007 Total Pixel Matrix Rows #2: 16457
---
> 0048,0006 Total Pixel Matrix Columns: 46000
> 0048,0007 Total Pixel Matrix Rows: 32914
```

The closest analogy I could think of would be TIFF files using private tags. Currently, a TIFF-based reader might elect to read this data and add it to the original metadata table and/or use it to interpret the data structure but there is no expectation that the reader will parse and read such private tags by default. 
From my perspective, the last commit is a similar interpretation of DICOM private data. In this case, the metadata from a private sequence excluded when reading the pixel data is also excluded from the original metadata table. It would be useful to get additional reviews and make sure we are happy with the behavior.

dgault reviewed (approved):
Tested using the sample files shared in `inbox/gh-4165` and confirmed that the issue with the 'scrambled' images has been resolved for all resolutions with this PR. All other builds and tests have been green with this PR included.

As for handling the metadata , based on the original description in https://github.com/ome/bioformats/issues/4165, I would agree with Seb's private tag analogy.  In this scenario leaving the tags related to the hidden pyramids out of the original metadata seems like the correct course of action from my side.
